compgen.yaml

```diff
@@ -0,0 +1,29 @@
+groups:
+  coremodel:
+    must:
+      # Requires that each coremodel component define a struct named "Coremodel"
+      # that implements the generic interface defined by the coremodel
+      # framework.
+      DefinesCoremodel:
+        ruletype: goident
+        name: Coremodel
+        kind: struct
+        implements:
+          - "github.com/grafana/grafana/internal/framework/coremodel.Interface"
+      # Requires that each coremodel component define a standard Thema
+      # LineageFactory func.
+      OffersLineageFactory:
+        ruletype: goident
+        name: Lineage
+        kind: func
+        implements:
+          - "github.com/grafana/thema.LineageFactory"
+      # Requires that each coremodel component define a func, usable by wire,
+      # that makes the coremodel available for use by things depending on the
+      # coremodel framework.
+      ProvidesCoremodel:
+        ruletype: goident
+        name: ProvideCoremodel
+        kind: func
+      #   isa:
+      #     - "github.com/grafana/grafana/internal/components.CoremodelProvider"
```

Comment:
Yeah this is all underdocumented, having come out of [a hackathon project](https://github.com/grafana/compgen).

The compgen.yaml file is one of those root-specifying files, like `go.mod` or `.git` - its position determines the fs tree searched by the compgen tool when looking for components. The tool is written generically, with the idea that we can introduce more such rules over time.

fs-based rules that chop up large codebases into zoned components with rules like this are very common in larger build system tools (e.g. bazel/buck/pants). `compgen` is an attempt to slice off one, narrow goal of such systems, UNIX philosophy-style, from those massive behemoths in order to lend more rigorous structure to our codebase.

i can remove the file (and the corresponding `componentroot.yaml` until we've at least gotten a README written, then re-add in a later PR, along with a new CI job to run `compgen validate`.

Reply:
A little more context...

Because we've exposed the system [as a library](https://github.com/grafana/compgen/tree/main/pkg/component), as well, it'll allow us to abstract the hardcoding of paths in things like [this command](https://github.com/grafana/grafana/blob/main/pkg/cmd/grafana-cli/commands/cuetsify_command.go) to instead ask compgen to tell us what the components are, what their paths are, and what properties the code at those paths has, and then write code on the basis of those properties.

So like, [this garbage](https://github.com/grafana/grafana/blob/main/pkg/cmd/grafana-cli/commands/cuetsify_command.go#L107-L113) and [this garbage](https://github.com/grafana/grafana/blob/main/pkg/cmd/grafana-cli/commands/cuetsify_command.go#L188) can go away, and instead be written dynamically against the verified properties of these named component groups. (Because yes, we can make more component groups as needed, e.g. for panel plugins, ds plugins)

The lack of an abstraction such as this was bugging the hell out of me last year - i wanted to avoid writing, explaining, and maintaining duplicate code and docs for things like applying the special rules for cuetsifying panel plugin `models.cue` files for both in-tree and third-party plugins.